### ISSUE TITLE
400 should be treated as conflict error when updating IAM policies

### ISSUE THREAD

**Issue by patrickdillon** [hashicorp/terraform-provider-google#8280]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version
Our project uses 0.13.4. This bug is in master & all versions of this provider.

### Affected Resource(s)

* google_iam

### Terraform Configuration Files

```tf
resource "google_service_account" "worker-node-sa" {
  account_id   = "${var.cluster_id}-w"
  display_name = "${var.cluster_id}-worker-node"
}

resource "google_project_iam_member" "worker-compute-viewer" {
  role   = "roles/compute.viewer"
  member = "serviceAccount:${google_service_account.worker-node-sa.email}"
}

resource "google_project_iam_member" "worker-storage-admin" {
  role   = "roles/storage.admin"
  member = "serviceAccount:${google_service_account.worker-node-sa.email}"
}
```

### Debug Output

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/release-openshift-ocp-installer-e2e-gcp-4.6/1325830980283404288/artifacts/e2e-gcp/installer/.openshift_install.log

To find the error in this massive log, you search for `Error 400: Service account`. Error is:

```
time="2020-11-09T16:04:31Z" level=debug msg="2020/11/09 16:04:31 [ERROR] module.iam: eval: *terraform.EvalSequence, err: Request \"Create IAM Members roles/compute.viewer serviceAccount:ci-op-f9p4hy7m-ad214-g4lgc-w@openshift-gce-devel-ci.iam.gserviceaccount.com for \\\"project \\\\\\\"openshift-gce-devel-ci\\\\\\\"\\\"\" returned error: Error applying IAM policy for project \"openshift-gce-devel-ci\": Error setting IAM policy for project \"openshift-gce-devel-ci\": googleapi: Error 400: Service account ci-op-7ijd7tr8-f92fc-j2vtt-m@openshift-gce-devel-ci.iam.gserviceaccount.com does not exist., badRequest"
```

### Expected Behavior

The provider should have retried creating applying the IAM policy, because there was a conflicting edit from another process (a service account was deleted).

### Actual Behavior

An error was triggered.

### Steps to Reproduce

1. Assign a new role to a service account
2. While the role is being applied, delete another service account.
This happens frequently in our CI environment we use service accounts at a large scale, so I doubt this is a commonly encountered issue.

### Important Factoids

This happens, in our case, when a service account is deleted by another process in between the two steps at the beginning and end of this code block: https://github.com/hashicorp/terraform-provider-google/blob/master/google/iam.go#L85-L101

I will open a PR with a suggested fix.

### References

Red Hat Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1896218#c18

**Comment by rileykarson** [comment]
@patrickdillon: When you refer to creating/deleting a service account, you mean the creation+deletion of the underlying SA and not applying policies to it, right? I'm confused about the circumstances here, how an issue encountered at service account creation/deletion time is resolved by changes to `iam.go`/IAM retry logic.

Is the problem that processing the deletion of a service account adds a delay to the creation of your new service account, adding a delay before it's available to link a policy to?

**Comment by patrickdillon** [comment]
> @patrickdillon: When you refer to creating/deleting a service account, you mean the creation+deletion of the underlying SA and not applying policies to it, right? 

No, sorry, I was unclear in my original message (now edited). I mean the application of the IAM roles to a service account. The deletion aspect is related (I believe) to the deletion of the actual service account.


> I'm confused about the circumstances here, how an issue encountered at service account creation/deletion time is resolved by changes to `iam.go`/IAM retry logic.
> Is the problem that processing the deletion of a service account adds a delay to the creation of your new service account, adding a delay before it's available to link a policy to?

Let me try to explain again: when applying a change to the IAM policy in this code: https://github.com/hashicorp/terraform-provider-google/blob/master/google/iam.go#L85-L101 there are essentially two steps: 1) read the existing policy 2) apply the updated policy. If a service account from the policy is deleted after step 1 but before step 2, the API responds with a 400 error, but the issue is that the policy pulled in step 1 is stale because there was a conflicting write(or delete, if you will). If we retried, we should get the correct policy.

Let me try to give an example. Let's say we want to add a membership role for `alice@acme.iam.gserviceaccount.com` and there is already a role for `bob@acme.iam.gserviceaccount.com`:

1. The existing policy is pulled (code I linked above) with `bob@acme.iam.gserviceaccount.com`
2. Another process deletes `bob@acme.iam.gserviceaccount.com` it is now something like `deleted:serviceAccount:bob@acme.iam.gserviceaccount.com`. The policy from step 1 is stale
3. `alice@acme.iam.gserviceaccount.com` is added to the stale policy along with `bob@acme.iam.gserviceaccount.com`
4. The API responds with 400 `bob@acme.iam.gserviceaccount.com does not exist` (because it is now  `deleted:serviceAccount:bob@acme.iam.gserviceaccount.com`)

**Comment by patrickdillon** [comment]
Apparently we are deleting the policy bindings after deleting the service account in our project. Reversing that order should mitigate the problem for us. Happy to close this, but perhaps accounting for this scenario would make your provider more robust.
